README.md
# Alaska Airlines Terraform Module Generator

[![Build Status](https://dev.azure.com/itsals/AlaskaAirlines_GitHub/_apis/build/status/Merge%20to%20master%20generator-aag-terraform-module?branchName=master&stageName=Publish)](https://dev.azure.com/itsals/AlaskaAirlines_GitHub/_build/latest?definitionId=6070&branchName=master) [![npm version](https://badge.fury.io/js/%40alaskaairgroup%2Fgenerator-aag-terraform-module.svg)](https://badge.fury.io/js/%40alaskaairgroup%2Fgenerator-aag-terraform-module)

A [Yeoman Generator](http://yeoman.io/) to create new [Terraform module](https://www.terraform.io/docs/configuration/modules.html) projects for Alaska Airlines. This generator creates a base repository to start a new terraform module along with [Terratest](https://github.com/gruntwork-io/terratest) or [kitchen-terraform](https://rubygems.org/gems/kitchen-terraform) to automatically test the module.

To stay up to date on our latest changes, visit our [Changelog](./docs/CHANGELOG.md)

## Install

To install the generator, run the following commands

```sh
npm install --global yo
npm install --global @alaskaairgroup/generator-aag-terraform-module
```

## Usage

To run the generator use the command below and provide your new module name for the prompt

```sh
yo @alaskaairgroup/generator-aag-terraform-module
```

```text
? Enter name for the new terraform module :  example-module
? Enter description for the new terraform module :  Example terraform module
? Enter author name :  sudokar
? Choose terraform version (Use arrow keys)
❯ 0.13
  0.11
? Choose test framework (Use arrow keys)
❯ Terratest
  kitchen-terraform
```

## Generated Terraform Module

### Terraform Files

- `main.tf` , `variables.tf` , `outputs.tf` files to module root path
- `examples` directory with module usage tf files
- `.terraform-version` file to module root path

### Testing Framework Files

- Option to choose test frameworks:
  - Terratest
  - kitchen-terraform
- `test` directory with an example test based on test framework selection

### Other Files

- `README.md`
- `Makefile`
- `.editorconfig`, `.gitignore` and `.gitattributes`
- `.pre-commit-config.yaml` for
  - `terraform fmt`
  - `terraform-docs`
  - `check-merge-conflict`
  - ( `go fmt`, `golint` ) or (`rubocop`)

## Post Generation

In the generated module directory, run the following command

```sh
make install
```

This will perform the following steps for you

- Initialize git repository
- Install pre-commit hooks
- Install Terraform
- Prepare testing framework

## Software Projects Included

- For module generator itself
  - [nodejs](https://nodejs.org/en/download/)
  - [Yeoman](https://yeoman.io/)
- For generated Terraform module
  - [homebrew](https://brew.sh/)
  - [terraform](https://learn.hashicorp.com/terraform/getting-started/install#installing-terraform)
  - [terraform-docs](https://github.com/segmentio/terraform-docs)
  - [pre-commit](https://pre-commit.com/#install)
- For generated Terraform module tests
  - **Terratest**
    - [golang](https://golang.org/doc/install#install)
    - [golint](https://github.com/golang/lint#installation)
    - [Terratest](https://github.com/gruntwork-io/terratest)
  - **kitchen-terraform**
    - [Chef Workstation](https://downloads.chef.io/chef-workstation/)
    - [kitchen-terraform](https://rubygems.org/gems/kitchen-terraform)

## Contributing

See [Contributing](./docs/CONTRIBUTING.md) for guidelines

### Building The Generator

If you are interested in contributing or modifying the generator itself, clone this repo and install all modules

```sh
npm install
```

### Testing The Generator

If you want to run the generator tests, you can run the following command

```sh
make test
```

### Using it in dev mode

In the `generator-aag-terraform-module` directory use the following command, this will link the local copy of the folder as a global node module

```sh
npm link
```

To revert the link operation

```sh
npm unlink
```

## Maintainers & Author

Maintainers: [Shadow Quests](shadowquests@alaskaair.com) (Alaska Airlines E-Commerce Platform Team)  
Original Author: [sudokar](https://github.com/sudokar)

## License

MIT


docs/CHANGELOG.md
# Semantic Release Automated Changelog

# [1.2.0](https://github.com/AlaskaAirlines/generator-aag-terraform-module/compare/v1.1.0...v1.2.0) (2020-11-09)


### Features

* Add support for Terraform 0.13 ([2b5dd42](https://github.com/AlaskaAirlines/generator-aag-terraform-module/commit/2b5dd426fdb8b5d780640b58abc6abbaa3876437))

# [1.1.0](https://github.com/AlaskaAirlines/generator-aag-terraform-module/compare/v1.0.0...v1.1.0) (2020-07-17)


### Features

* Add Azure DevOps Pipeline YAML ([7eaf83d](https://github.com/AlaskaAirlines/generator-aag-terraform-module/commit/7eaf83d677c5953404d0ffc57e6600c63a05047e))

# 1.0.0 (2020-07-14)


### Bug Fixes

* fix for azure-pipelines.yml ([#4](https://github.com/AlaskaAirlines/generator-aag-terraform-module/issues/4)) ([ea8e54b](https://github.com/AlaskaAirlines/generator-aag-terraform-module/commit/ea8e54bc014ee564d537a91e7d3c7f93072aa7a8))
* Update Azure Pipelines YAML ([240b9c2](https://github.com/AlaskaAirlines/generator-aag-terraform-module/commit/240b9c25c4cdc1e38483e65762146f7110914a84))


generators/app/templates/test/terratest/example_test.go
package test

import (
	"testing"

	"github.com/gruntwork-io/terratest/modules/terraform"
	"github.com/stretchr/testify/assert"
)

func TestTerraformExample(t *testing.T) {
	// Arrange
	terraformOptions := &terraform.Options{
		TerraformDir: "../examples/.",
	}
	defer terraform.Destroy(t, terraformOptions)

	// Act
	terraform.InitAndApply(t, terraformOptions)

	// Assert
	assert := assert.New(t)

	outputValue := terraform.Output(t, terraformOptions, "output_name")
	assert.NotNil(outputValue)
	assert.Equal("output_value", outputValue)
}


generators/app/index.js
const Generator = require('yeoman-generator');
const yosay = require('yosay');

module.exports = class extends Generator {
  async prompting() {
    this.log(yosay('Welcome to the aag-terraform-module generator!'));

    this.answers = await this.prompt([
      {
        type: 'input',
        name: 'name',
        message: 'Enter name for the new terraform module : ',
        store: true,
      },
      {
        type: 'input',
        name: 'description',
        message: 'Enter description for the new terraform module : ',
        store: true,
      },
      {
        type: 'input',
        name: 'author',
        message: 'Enter author name : ',
        default: 'Shadow Quests (E-Commerce Platform Team) <shadowquests@alaskaair.com>',
        store: true,
      },
      {
        type: 'list',
        name: 'tfVersion',
        message: 'Choose terraform version',
        store: true,
        choices: [
          {
            name: '0.13',
            value: '13',
          },
          {
            name: '0.11',
            value: '11',
          },
        ],
      },
      {
        type: 'list',
        name: 'testFramework',
        message: 'Choose test framework',
        store: true,
        choices: [
          {
            name: 'Terratest',
            value: '1',
          },
          {
            name: 'kitchen-terraform',
            value: '2',
          },
        ],
      },
    ]);
  }

  writing() {
    this.destinationRoot(this.answers.name);
    this.fs.copyTpl(
      `${this.templatePath()}/.!(gitignorefile|gitattributesfile|pre-commit-config|terraform-version)*`,
      this.destinationRoot(),
      {},
    );

    this.fs.copyTpl(
      this.templatePath('.gitignorefile'),
      this.destinationPath('.gitignore'),
      {
        testFramework: this.answers.testFramework,
      },
    );

    this.fs.copyTpl(
      this.templatePath('.gitattributesfile'),
      this.destinationPath('.gitattributes'),
      {
        testFramework: this.answers.testFramework,
      },
    );

    this.fs.copyTpl(
      this.templatePath('.pre-commit-config.yaml'),
      this.destinationPath('.pre-commit-config.yaml'),
      {
        testFramework: this.answers.testFramework,
      },
    );

    this.fs.copyTpl(
      this.templatePath('.terraform-version'),
      this.destinationPath('.terraform-version'),
      {
        tfVersion: this.answers.tfVersion,
      },
    );

    this.fs.copyTpl(
      `${this.templatePath()}/(Brewfile|Makefile)`,
      this.destinationRoot(),
      {
        testFramework: this.answers.testFramework,
      },
    );

    this.fs.copyTpl(
      `${this.templatePath()}/docs/*.md`,
      `${this.destinationPath()}/docs`,
    );

    this.fs.copyTpl(
      `${this.templatePath()}/**/*.tf`,
      this.destinationRoot(),
      {
        tfVersion: this.answers.tfVersion,
      },
    );

    this.fs.copyTpl(
      `${this.templatePath()}/examples/*.example`,
      `${this.destinationPath()}/examples`,
    );

    this.fs.copyTpl(
      this.templatePath('azure-pipelines.yml'),
      this.destinationPath('azure-pipelines.yml'),
      {
        testFramework: this.answers.testFramework,
      },
    );

    if (this.answers.testFramework === '1') {
      this.fs.copyTpl(
        `${this.templatePath()}/test/terratest/*.go`,
        `${this.destinationRoot()}/test`,
        {},
      );
    } else {
      this.fs.copyTpl(
        `${this.templatePath()}/test/kitchen-terraform/.*`,
        this.destinationRoot(),
        {},
      );
      this.fs.copyTpl(
        `${this.templatePath()}/test/kitchen-terraform/**/*`,
        this.destinationRoot(),
        {},
      );
    }

    this.fs.copyTpl(
      this.templatePath('_.releaserc'),
      this.destinationPath('.releaserc'), {
        name: this.answers.name,
      },
    );

    this.fs.copyTpl(
      this.templatePath('_package.json'),
      this.destinationPath('package.json'), {
        name: this.answers.name,
        author: this.answers.author,
      },
    );

    this.fs.copyTpl(
      this.templatePath('_README.md'),
      this.destinationPath('README.md'),
      {
        name: this.answers.name,
        description: this.answers.description,
        author: this.answers.author,
        testFramework: this.answers.testFramework,
      },
    );
  }
};
